content/blog/2025-12-16-ChocoPyToBril

```diff
@@ -0,0 +1,110 @@
+          "type": "int",
+          "args": [
+            "t2",
+            "t1"
+          ]
+        },
+        {
+          "op": "sub",
+          "dest": "t4",
+          "type": "int",
+          "args": [
+            "t0",
+            "t3"
+          ]
+        },
+        {
+          "op": "id",
+          "dest": "x",
+          "type": "int",
+          "args": [
+            "t4"
+          ]
+        },
+        {
+          "op": "print",
+          "args": [
+            "x"
+          ]
+        },
+        {
+          "op": "ret"
+        }
+      ]
+    }
+  ]
+}
+Which after using bril2txt converts to the following:
+@main {
+  t0: int = const 1;
+  t1: int = const 2;
+  t2: int = div t0 t1;
+  t3: int = mul t2 t1;
+  t4: int = sub t0 t3;
+  x: int = id t4;
+  print x;
+  ret;
+}
+
+### What were the hardest parts to get right?
+The hardest parts for me definitely had to be function definition and while loops. I struggled for a bit to understand completely how functions were placed in bril JSON, as well as how to translate them correctly. The reason I was able to implement both was taking a whiteboard and breaking both of them down structurally, and then piecing them together in my code one attribute by attribute. In particular, one menacing error that haunted me throughout this project was that "1 function required". It took me several hours to debug this error as it was not alot to go off of. After paintstakingly comparing bril JSONs to the benchmarks in the bril repo, I realized that my error lied in how I had structured my call op, where it had the func attribute instead of funcs, and had an array. I was confused on why it had multiple funcs as I have never seen a bril file that actually makes the json use more than one function within the call, but after fixing it all was well.
+
+### Were you successful?
+I was overall successful as all my test cases passed. For my test cases I used all the test cases that would work from "https://github.com/gabe-raulet/chocopy2bril/tree/master/examples", as well as base test cases that were used to work out and develop my solution. The tests did have to be rewritten slightly to be checked (where I would remove main input), but other than that they were the same as the original. I made 20 tests, and they all had the exact same output as they would in python. Originally, in my proposal I had stated 20 tests from a directory. However, I realized, funnily enough, that this directory didn’t have 20 test cases and that the amount of files I saw in the directory was inflated due to .opt.bril, _hand_impl, and .compiled.opt files. Due to this, I fetched more test cases from online ("https://github.com/yangdanny97/chocopy-python-compiler/tree/master/tests/"). In order to test, I checked the difference in output between executing with CPython and executing with my compiler implementation. There is a run_diff.sh script that executes this for every test in the test folder.
```

Comment:
> "https://github.com/gabe-raulet/chocopy2bril/tree/master/examples"

Please use an actual Markdown link. The same for the other URL in this paragraph.